scipy/optimize/shgo_m/sobol_seq.py

```diff
@@ -0,0 +1,365 @@
+      |   14 |       1110 |   1
+      |   15 |       1111 |   5
+      |   16 |      10000 |   1
+      |   17 |      10001 |   2
+      | 1023 | 1111111111 |   1
+      | 1024 | 0000000000 |   1
+      | 1025 | 0000000001 |   1
+
+    Parameters:
+      Input, integer N, the integer to be measured.
+      N should be nonnegative.
+
+      Output, integer BIT, the position of the low 1 bit.
+    """
+    bit = 1
+    i = np.floor(n)
+    while i != 2 * (i // 2):
+        bit += 1
+        i //= 2
+    return bit
+
+
+def i4_sobol_generate(dim_num, n, skip=1):
+    """
+    i4_sobol_generate generates a Sobol dataset.
+
+    Parameters:
+      Input, integer dim_num, the spatial dimension.
+      Input, integer N, the number of points to generate.
+      Input, integer SKIP, the number of initial points to skip.
+
+      Output, real R(M,N), the points.
+    """
+    r = np.full((n, dim_num), np.nan)
+    for j in range(n):
+        seed = j + skip
+        r[j, 0:dim_num], next_seed = i4_sobol(dim_num, seed)
+
+    return r
+
+
+def i4_sobol(dim_num, seed):
+    """
+    i4_sobol generates a new quasirandom Sobol vector with each call.
+
+    Discussion:
+      The routine adapts the ideas of Antonov and Saleev.
+
+    Reference:
+      Antonov, Saleev,
+      USSR Computational Mathematics and Mathematical Physics,
+      Volume 19, 1980, pages 252 - 256.
+
+      Paul Bratley, Bennett Fox,
+      Algorithm 659:
+      Implementing Sobol's Quasirandom Sequence Generator,
+      ACM Transactions on Mathematical Software,
+      Volume 14, Number 1, pages 88-100, 1988.
+
+      Bennett Fox,
+      Algorithm 647:
+      Implementation and Relative Efficiency of Quasirandom
+      Sequence Generators,
+      ACM Transactions on Mathematical Software,
+      Volume 12, Number 4, pages 362-376, 1986.
+
+      Ilya Sobol,
+      USSR Computational Mathematics and Mathematical Physics,
+      Volume 16, pages 236-242, 1977.
+
+      Ilya Sobol, Levitan,
+      The Production of Points Uniformly Distributed in a Multidimensional
+      Cube (in Russian),
+      Preprint IPM Akad. Nauk SSSR,
+      Number 40, Moscow 1976.
+
+    Parameters:
+      Input, integer DIM_NUM, the number of spatial dimensions.
+      DIM_NUM must satisfy 1 <= DIM_NUM <= 40.
+
+      Input/output, integer SEED, the "seed" for the sequence.
+      This is essentially the index in the sequence of the quasirandom
+      value to be generated.  On output, SEED has been set to the
+      appropriate next value, usually simply SEED+1.
+      If SEED is less than 0 on input, it is treated as though it were 0.
+      An input value of 0 requests the first (0-th) element of the sequence.
+
+      Output, real QUASI(DIM_NUM), the next quasirandom vector.
+    """
+    global atmost
+    global dim_max
+    global dim_num_save
+    global initialized
+    global lastq
+    global log_max
+    global maxcol
+    global poly
+    global recipd
+    global seed_save
+    global v
```

Comment:
This needs to be refactored into a class so that we can have several instances alive that don't interact with each other.

Reply:
Refactored in 8d6d537.